### PR TITLE
Add base feature pyramid network (FPN) to Keras CV

### PR DESCRIPTION
**Pull request by qlzh727**
# What does this PR do?
Add base feature pyramid network (FPN) to Keras CV

This is a update for https://github.com/keras-team/keras-cv/pull/504 with more alignment to Keras APIs.

Next step:
1. Add a FPN subclass with extension to retina-net, which has extra outputs above the
   backbone feature levels.
2. Migrate the existing retina-net FPN to use the new implementation.
3. Change existing backbones to emit the output features for each of the feature level.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

### CODE REVIEW

**Review by LukeWood — CHANGES_REQUESTED** [review]
Super minor changes.  Once submitted I’m happy to retire the internal FPN and use this in the RetinaNet - so let’s not worry too much about the FPN in internal

File: keras_cv/layers/feature_pyramid.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+
+# TODO(scottzhu): Register it later due to the conflict in the retina_net
+# @tf.keras.utils.register_keras_serializable(package="keras_cv")
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/feature_pyramid.py:19]: (Or I can do this in the PR that i will create to migrate to this in the RetinaNet)

File: keras_cv/layers/feature_pyramid.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+    The inputs to the layer should be a dict with int keys should match the
+    pyramid_levels, e.g. for `pyramid_levels` = [2,3,4,5], the expected input dict should
+    be `{2:c2, 3:c3, 4:c4, 5:c5}`.
+
+    The output of the layer will have same structures as the inputs, a dict with int keys
+    and value for each of the level.
+
+    Args:
+        pyramid_levels: a sorted python int list that specifies all the values of level
```
Comment (LukeWood) [inline keras_cv/layers/feature_pyramid.py:52]: Should we support ints for this argument and just perform range() to produce the keys?

File: keras_cv/layers/feature_pyramid.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        self.pyramid_levels = sorted(pyramid_levels)
+        self.num_channels = num_channels
+
+        # required for successful serialization
+        self.lateral_layers_passed = lateral_layers
+        self.output_layers_passed = output_layers
+        self.top_down_op_passed = top_down_op
+        self.merge_op_passed = merge_op
+
+        if not lateral_layers:
```
Comment (LukeWood) [inline keras_cv/layers/feature_pyramid.py:106]: lets refactor the creation of the defaults and these big if else statements into helper functions.

i.e.

```
self.lateral_layers = _parse_lateral_layers(lateral_layers)
```

File: keras_cv/layers/feature_pyramid.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                or sorted(lateral_layers.keys()) != self.pyramid_levels
+            ):
+                raise ValueError(
+                    f"Expect lateral_layers to be a dict with keys as "
+                    f"{self.pyramid_levels}, got {lateral_layers}"
+                )
+            self.lateral_layers = lateral_layers
+
+        # Output conv2d layers.
+        if not output_layers:
```
Comment (LukeWood) [inline keras_cv/layers/feature_pyramid.py:130]: same for output

File: keras_cv/layers/feature_pyramid.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                or sorted(output_layers.keys()) != self.pyramid_levels
+            ):
+                raise ValueError(
+                    f"Expect output_layers to be a dict with keys as "
+                    f"{self.pyramid_levels}, got {output_layers}"
+                )
+            self.output_layers = output_layers
+
+        # the same upsampling layer is used for all levels
+        if not top_down_op:
```
Comment (LukeWood) [inline keras_cv/layers/feature_pyramid.py:152]: and lastly I think just for consistency lets do the same with these

File: keras_cv/layers/feature_pyramid.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    def call(self, features):
+        # Note that this assertion might not be true for all the subclasses. It is
+        # possible to have FPN that has high levels than the height of backbone outputs.
+        if (
+            not isinstance(features, dict)
+            or sorted(features.keys()) != self.pyramid_levels
+        ):
+            raise ValueError(
+                "Expect the input features to be a dict with int keys that"
+                f"matches to the {self.pyramid_levels}, got: {features}"
```
Comment (LukeWood) [inline keras_cv/layers/feature_pyramid.py:171]: lets tell users where the self.pyramid_levels dict that will be formatted in is coming from.

FeaturePyramid expects input features to be a dict with int keys that match the values provided in `pyramid_levels`.  want keys(inputs)={self.pyramid_levels}, got={features.keys()}

File: keras_cv/layers/feature_pyramid.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            )
+        return self.build_feature_pyramid(features)
+
+    def build_feature_pyramid(self, input_features):
+        # To illustrate the connection/topology, the basic flow for a FPN with level
+        # 3, 4, 5 is like below:
+        #
+        # input_l5 -> conv2d_1x1_l5 ----V---> conv2d_3x3_l5 -> output_l5
```
Comment (LukeWood) [inline keras_cv/layers/feature_pyramid.py:179]: I like the diagram - nice!

File: keras_cv/layers/feature_pyramid.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                output = self.merge_op([output, upstream_output])
+            output_features[level] = output
+
+        # Post apply the output layers so that we don't leak them to the down stream level
+        for level in reversed_levels:
+            output_features[level] = self.output_layers[level](output_features[level])
+
+        return output_features
+
+    # TODO(scottzhu): Add serialization with get_config/from_config(). It might have
```
Comment (LukeWood) [inline keras_cv/layers/feature_pyramid.py:207]: As long as we save the original inputs in the constructor, I think get_config()/from_config() will actually automatically handle sub-layers as long as sub-layers are registered as serializables themselves

**Review by LukeWood — APPROVED** [review]
did a more thorough pass and I've provided some minor comments - but overall I think this PR looks really good.

Thanks!

**Review by LukeWood — CHANGES_REQUESTED** [review]
(Hit approve instead of request changes)

File: keras_cv/layers/feature_pyramid.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    This implements the paper:
+      Tsung-Yi Lin, Piotr Dollar, Ross Girshick, Kaiming He, Bharath Hariharan, and
+      Serge Belongie.
+      Feature Pyramid Networks for Object Detection.
+      (https://arxiv.org/pdf/1612.03144)
+
+    Feature Pyramid Networks (FPNs) are basic components that are added to an
+    existing CNN to extract features at different scales. For the basic FPN, the
+    inputs are features `Ci` from different levels of a CNN, where the feature is
+    scaled from the image size by `1/2^i` for any level `i`.
```
Comment (tanzhenyu) [inline keras_cv/layers/feature_pyramid.py:32]: where the feature map size is scaled from the image size by a factor of ...

File: keras_cv/layers/feature_pyramid.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+    This implements the paper:
+      Tsung-Yi Lin, Piotr Dollar, Ross Girshick, Kaiming He, Bharath Hariharan, and
+      Serge Belongie.
+      Feature Pyramid Networks for Object Detection.
+      (https://arxiv.org/pdf/1612.03144)
+
+    Feature Pyramid Networks (FPNs) are basic components that are added to an
+    existing CNN to extract features at different scales. For the basic FPN, the
+    inputs are features `Ci` from different levels of a CNN, where the feature is
```
Comment (tanzhenyu) [inline keras_cv/layers/feature_pyramid.py:31]: which is usually the last block for each level

File: keras_cv/layers/feature_pyramid.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    and value for each of the level.
+
+    Args:
+        pyramid_levels: a sorted python int list that specifies all the values of level
+            `i` for which the feature Ci will be specified.
+        num_channels: an integer representing the number of channels for the FPN
+            operations. Defaults to 256.
+        lateral_layers: a python dict with int keys that matches to each of the pyramid
+            level. The values of the dict should be `keras.Layer`, which will be called
+            with feature inputs from backbone at each level. Default to None, and a
```
Comment (tanzhenyu) [inline keras_cv/layers/feature_pyramid.py:58]: with feature activation outputs from backbone

File: keras_cv/layers/feature_pyramid.py (hunk 7 above)
Comment (tanzhenyu) [reply keras_cv/layers/feature_pyramid.py:179]: +100

File: keras_cv/layers/feature_pyramid.py [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+            `keras.Conv2D` layer with kernel 1x1 will be created for each pyramid level.
+        output_layers: a python dict with int keys that matches to each of the pyramid
```
Comment (tanzhenyu) [inline keras_cv/layers/feature_pyramid.py:60]: when does it need to be op and when does it need to be layer? (Asking this as users might get confused)

File: keras_cv/layers/feature_pyramid_test.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+        inputs = {2: c2, 3: c3, 4: c4, 5: c5}
+        layer = FeaturePyramid(pyramid_levels=[2, 3, 4, 5])
+        output = layer(inputs)
+        for level in inputs.keys():
+            self.assertEquals(output[level].shape[1], inputs[level].shape[1])
+            self.assertEquals(output[level].shape[2], inputs[level].shape[2])
+            self.assertEquals(output[level].shape[3], layer.num_channels)
+
+    def test_keras_tensor(self):
```
Comment (tanzhenyu) [inline keras_cv/layers/feature_pyramid_test.py:61]: better naming with this test?

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/layers/feature_pyramid.py (hunk 9 above)
Comment (qlzh727) [reply keras_cv/layers/feature_pyramid.py]: Done.

File: keras_cv/layers/feature_pyramid_test.py (hunk 13 above)
Comment (qlzh727) [reply keras_cv/layers/feature_pyramid_test.py]: Done.

**Review by tanzhenyu — APPROVED** [review]
Thanks for the PR!

**Pull request merged**